OWL Axiom support in stated taxonomy and ECL queries

cmark opened the pull request:
This PR adds support for OWL Axiom based stated form.

* Port stated ECL support from https://github.com/IHTSDO/snow-owl fork.
* Add `snomed-owl-toolkit` dependency (version 2.3.2).
* Add new `classAxiomRelationships` and `gciAxiomRelationships` fields to reference set members where the owlExpression field is populated with valid OWL Axiom and can be converted to relationship fragments.
* Use the newly added fields to query stated taxonomy for `statedParents` and `statedAncestors` updates.
* Use the newly added fields to query stated relationships in stated ECL evaluation.

cmark committed:
OWLAP-48: add snomed-owl-toolkit 2.3.1 dependency

cmark committed:
OWLAP-48 OWLAP-46: add additional and gciAxiomRelationships field...

...to SnomedRefSetMemberIndexEntry

Support MRCM_GROUPED property filter in SnomedRefSetMemberSearchRequest.
Add index tests for additionalAxiomRelationships.

cmark committed:
OWLAP-48 OWLAP-46: rename additionalAxioms to classAxioms

cmark committed:
OWLAP-48 OWLAP-46: update owl-toolkit version to 2.3.2

cmark committed:
OWLAP-48 OWLAP-46: add GCI axiom test case

cmark committed:
OWLAP-48 OWLAP-46: calculate class and gci axioms during commit

cmark committed:
OWLAP-48 OWLAP-46: calculate stated parents/ancestors long arrays...

...during commits from OWL axiom members.

Support both stated relationships and OWL class axioms when computing
stated tree changes during commit.
Remove unused SnomedInvalidTaxonomyMergeConflictRule.
Filter unsupported owlExpressions when converting them to axiom
relationships.

cmark committed:
OWLAP-48 OWLAP-46: port statedEclFilter changes from IHTSDO/snow-owl

cmark committed:
OWLAP-48 OWLAP-46: inactive members should not have axiom relationships

cmark committed:
OWLAP-48 OWLAP-46: commit progress on owl axiom stated tree support

cmark committed:
OWLAP-48 OWLAP-46: populate edges in old and new taxonomy graphs

cmark committed:
OWLAP-48 OWLAP-46: improve index queries to reduce memory pressure

Where possible use field select to load only the necessary amount of
data into memory.

cmark committed:
OWLAP-48 OWLAP-46: load existing axiom members during RF2 import

cmark committed:
OWLAP-48 OWLAP-46: add change processing tests for OWL Axiom members

cmark committed:
OWLAP-48 OWLAP-46: add direct executor to ExecutorServiceFactory

Use DIRECT executor in TestBranchContext.
Remove duplicate TestBranchContext class.

cmark committed:
OWLAP-48 OWLAP-46: fix branch context init in unit tests

cmark committed:
OWLAP-48 OWLAP-46: return empty axiom statements when conversion fails

cmark committed:
OWLAP-48 OWLAP-46: fix incorrect field selection in ComponentLookup

cmark committed:
OWLAP-48 OWLAP-46: add RF2TaxonomyGraph init method implementation

cmark committed:
OWLAP-48 OWLAP-46: fix failed snomed API test cases

cmark committed:
OWLAP-48 OWLAP-46: support axioms in stated ECL queries

cmark committed:
Merge branch '6.x' into...

...feature/OWLAP-46-48-owl-axiom-relationship-support

Conflicts:
	pom.xml
	snomed/com.b2international.snowowl.snomed.api.rest.tests/.launch/snomed-e2e-tests.launch
	snomed/com.b2international.snowowl.snomed.datastore.tests/.launch/snomed-core-unit-tests.launch
	snomed/com.b2international.snowowl.snomed.datastore/META-INF/MANIFEST.MF

apeteri commented on net4j/com.b2international.snowowl.eventbus/src/com/b2international/snowowl/eventbus/EventBusUtil.java:
```diff
@@ -86,7 +87,16 @@ public void handle(IMessage message) {
+	 * @return an {@link EventBus} with the specified description and 1 direct thread worker.
+	 */
+	public static IEventBus getDirectBus(String name) {
+		final IEventBus bus = new EventBus(name, 1, ExecutorServiceFactory.DIRECT);
```
What is the difference between `name, 1, ExecutorServiceFactory.DIRECT` and `name, 0` as arguments?

apeteri commented on snomed/com.b2international.snowowl.snomed.datastore/src/com/b2international/snowowl/snomed/core/ecl/SnomedEclRefinementEvaluator.java:
```diff
@@ -93,17 +106,20 @@
 
-	static final Set<String> ALLOWED_CHARACTERISTIC_TYPES = ImmutableSet.of(Concepts.INFERRED_RELATIONSHIP, Concepts.ADDITIONAL_RELATIONSHIP);
+	static final Set<String> STATED_CHARACTERISTIC_TYPES = ImmutableSet.of(Concepts.STATED_RELATIONSHIP, Concepts.ADDITIONAL_RELATIONSHIP);
+	static final Set<String> INFERRED_CHARACTERISTIC_TYPES = ImmutableSet.of(Concepts.INFERRED_RELATIONSHIP, Concepts.ADDITIONAL_RELATIONSHIP);
```
Do we want to make the distinction between grouped and ungrouped additional properties in the inferred form?

apeteri commented on snomed/com.b2international.snowowl.snomed.datastore/src/com/b2international/snowowl/snomed/core/ecl/SnomedEclRefinementEvaluator.java:
```diff
@@ -564,36 +579,121 @@ public SnomedEclRefinementEvaluator(EclExpression focusConcepts) {
+			final Set<Property> axiomStatements = evalAxiomStatements(context, groupedRelationshipsOnly, sourceIds, typeIds, destinationIds);
+			return then.then(relationshipStatements -> ImmutableSet.<Property>builder().addAll(relationshipStatements).addAll(axiomStatements).build());
+		} else {
+			return then;
```
Please find another name for this local variable.

apeteri commented on snomed/com.b2international.snowowl.snomed.datastore/src/com/b2international/snowowl/snomed/core/ecl/SnomedEclRefinementEvaluator.java:
```diff
@@ -564,36 +579,121 @@ public SnomedEclRefinementEvaluator(EclExpression focusConcepts) {
+				activeOwlAxiomMemberQuery.filter(SnomedRefSetMemberIndexEntry.Expressions.referencedComponentIds(sourceIds));
+			}
+			
+			final Query<SnomedRefSetMemberIndexEntry> activeAxiomISARelationshipsQuery = Query.select(SnomedRefSetMemberIndexEntry.class)
```
⚠️ This doesn't seem to be an IS A-only axiom query, there is a `typeId` filter above.

cmark replied:
Renamed!

apeteri commented on snomed/com.b2international.snowowl.snomed.datastore/src/com/b2international/snowowl/snomed/datastore/request/SnomedConceptSearchRequestBuilder.java:
```diff
@@ -106,6 +106,18 @@ public final SnomedConceptSearchRequestBuilder filterByEcl(String expression) {
+	/**
+	 * Filter matches in the stated tree by the specified Expression Constraint Language (ECL) expression. 
+	 * The currently supported ECL version is v1.3. See <a href="http://snomed.org/ecl">ECL Specification and Guide</a> or
+	 * <a href="http://www.snomed.org/news-articles/expression-constraint-language">About ECL</a> for more information.
```
This link is non-reachable ⚡️

cmark replied:
Fixed!

apeteri commented on snomed/com.b2international.snowowl.snomed.datastore/src/com/b2international/snowowl/snomed/datastore/taxonomy/Edges.java:
```diff
@@ -0,0 +1,51 @@
+	
+	@Override
+	public int hashCode() {
+		return Objects.hash(sourceId, destinationIds);
```
`destinationIds` also needs to be hashed by content:

```suggestion
		return Objects.hash(sourceId, Arrays.hashCode(destinationIds));
```

cmark replied:
Fixed!

apeteri commented on snomed/com.b2international.snowowl.snomed.datastore/src/com/b2international/snowowl/snomed/datastore/taxonomy/Taxonomies.java:
```diff
@@ -48,62 +57,141 @@
+					.filter(Expressions.nestedMatch(SnomedRefSetMemberIndexEntry.Fields.CLASS_AXIOM_RELATIONSHIP, 
+						Expressions.builder()
+							.filter(typeId(Concepts.IS_A))
+							.filter(destinationIds(concepts))
```
⚠️ For stated IS A relationships above, the destination is not restricted unless `filterByConceptIds` is enabled.

cmark replied:
Right, excluded `destinationIds` filter in that case.

apeteri commented on snomed/com.b2international.snowowl.snomed.datastore/src/com/b2international/snowowl/snomed/datastore/taxonomy/TaxonomyGraphUpdater.java:
```diff
@@ -211,97 +218,55 @@ public void run() {
+					.map(SnomedOWLRelationshipDocument::getDestinationId)
+					.mapToLong(Long::parseLong)
+					.toArray();
+				graphToUpdate.addEdge(owlMember.getUuid(), Long.parseLong(owlMember.getReferencedComponentId()), destinationIds);
```
Are OWL expression destination changes also handled in `addEdge`?

cmark replied:
Yes!

cmark committed:
Merge branch 'origin/6.x' into...

...feature/OWLAP-46-48-owl-axiom-relationship-support

Conflicts:
	pom.xml

cmark committed:
OWLAP-48 OWLAP-46: fix requested changes

Use simple Map instead of primitive IntMap in TaxonomyGraph.
Renamed incorrectly named variables.
Remove no longer valid ECL documentation web links.
Fix OWL Axiom based graph initialization during RF2 import.

cmark committed:
OWLAP-48 OWLAP-46: make defensive copy of edgeId sets in Taxonomy
